core/arxiv/submission/domain/event/versioning/__init__.py

```diff
@@ -0,0 +1,131 @@
+to database migrations, except that the "meat" of the event data are dicts
+stored as JSON and thus ALTER commands won't get us all that far.
+
+Writing version mappings
+========================
+Any new version of this software that includes changes to existing
+event/command classes that would break events from earlier versions **MUST**
+include a version mapping module. The module should include a mapping class
+(a subclass of :class:`.BaseVersionMapping`) for each event type for which
+there are relevant changes.
+
+See :mod:`.versioning.version_0_0_0_example` for an example.
+
+Each such class must include an internal ``Meta`` class with its software
+version and the name of the event type to which it applies. For example:
+
+.. code-block:: python
+
+   from ._base import BaseVersionMapping
+
+   class SetAbstractMigration(BaseVersionMapping):
+       class Meta:
+           event_version = "0.2.12"     # Must be a semver.
+           event_type = "SetAbstract"
+
+
+In addition, it's a good idea to include some test data that can be used to
+verify the behavior of the migration. You can do this by adding a ``tests``
+attribute to ``Meta`` that includes tuples of the form
+``(original: EventData, expected: EventData)``. For example:
+
+
+.. code-block:: python
+
+   from ._base import BaseVersionMapping
+
+   class SetAbstractMigration(BaseVersionMapping):
+       class Meta:
+           event_version = "0.2.12"     # Must be a semver.
+           event_type = "SetAbstract"
+           tests = [({"event_version": "0.2.11", "abstract": "very abstract"},
+                     {"event_version": "0.2.12", "abstract": "more abstract"})]
+
+
+Transformation logic can be implemented for individual fields, or for the event
+datum as a whole.
+
+Transforming individual fields
+------------------------------
+Transformers for individual fields may be implemented by
+defining instance methods with the name ``transform_{field}`` and the signature
+``(self, original: EventData, key: str, value: Any) -> Tuple[str, Any]``.
+The return value is the field name and transformed value. Note that the field
+name may be altered here, and the original field name will be omitted from the
+final transformed representation of the event datum.
+
+Transforming the datum as a whole
+---------------------------------
+A transformer for the datum as a whole may be implemented by defining an
+instance method named ``transform`` with the signature
+``(self, original: EventData, transformed: EventData) -> EventData``. This is
+called **after** the transformers for individual fields; the second positional
+argument is the state of the datum at that point, and the first positional
+argument is the state of the datum before transformations were applied.
+"""
+
+import copy
+from ._base import EventData, BaseVersionMapping, Version
+
+from arxiv.base.globals import get_application_config
+
+
+def map_to_version(original: EventData, target: str) -> EventData:
```

Comment:
So this function is really the key to the PR in some sense - when calling code post-migration, you're going to have `map_to_version` which will update all events as necessary.

Is there some better way to set `target` on an application configuration level so that each event-retrieval doesn't have to be wrapped in `map_to_version` manually?

Reply:
Looking at the example now, is the intended use case to run a migration script once, rather than on-the-fly retrieval of data?

Reply:
Aha! The key to this is a bit buried. Indeed, per docs, this is on-the-fly only. This function gets called in the ``event_factory``, which is used to reconstruct ``Event`` instances from data.

https://github.com/arXiv/arxiv-submission-core/pull/54/files#diff-3a6ea2cbb230676b45943a23008e598bR1200